### PR TITLE
(types): improve types for setState

### PR DESCRIPTION
**Pull request by TkDodo**
@dai-shi I think this would be a minimal invasive fix for #317 . I can't know for sure until I write some type tests, which is why this is a draft so far :) What do you think?

### CODE REVIEW

File: src/vanilla.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 export type State = Record<string | number | symbol, unknown>
-export type PartialState<T extends State> =
-  | Partial<T>
-  | ((state: T) => Partial<T>)
+export type PartialState<T extends State, K extends keyof T = keyof T> =
+  | (Pick<T, K> | T)
+  | ((state: T) => Pick<T, K> | T)
```
Comment (dai-shi) [inline src/vanilla.ts:4]: Do you know if we really need `| T`?

**Review by dai-shi — COMMENTED** [review]
Nice try. Please keep going.

File: src/vanilla.ts (hunk 1 above)
Comment (TkDodo) [reply src/vanilla.ts:4]: I do not :)

I got this from the react typings on `setState` of a class component:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/6c49e45842358ba59a508e13130791989911430d/types/react/v16/index.d.ts#L489-L495

Judging from the comment, it is needed for intelliSense 😅

Comment (dai-shi) [reply src/vanilla.ts:4]: Ah ha. We need to put the link in the code comment, then.

Comment (TkDodo) [reply src/vanilla.ts:6]: done here: b142dcda

**Comment by TkDodo** [comment]
@dai-shi I added some tests and tried it out in of my projects where we use zustand, and it worked pretty well. Please have another look

**Comment by TkDodo** [comment]
when I'm trying to run my example in the codesandbox build, I am getting this:

Screenshot 2021-02-27 at 18 19 17

I am not getting this in other PRs 🤔

**Comment by TkDodo** [comment]
okay, this might be something with codesandbox. I installed:

`"zustand": "https://pkg.csb.dev/pmndrs/zustand/commit/5564b272/zustand"`

locally and all was fine, the example now errors on `{ count: undefined }` as expected :)

**Comment by dai-shi** [comment]
(yeah, sometimes I get similar weird behaviors in codesandbox, like caching issue.)

**Comment by dai-shi** [comment]
Confirmed locally that this also shows an error as expected:
```ts
import create from "zustand";
import { combine } from "zustand/middleware";

export const useStore = create(
  combine(
    {
      count: 0
    },
    (set) => ({
      inc: () => set({ count: undefined })
    })
  )
);
```

**Review by dai-shi — APPROVED** [review]

**Pull request merged**